File: Cpp/Example7-3-DemModel/Example7-3-DemModel.cpp
#include <gdal_priv.h>

#include <fstream>
#include <iomanip>
#include <iostream>
#include <nlohmann/json.hpp>

using namespace std;
using namespace nlohmann;

size_t pointNum = 0;
size_t binBufNum = 0;
size_t indicesNum = 0;

void CreateBinFile() {
  string workDir = getenv("GISBasic");
  string demPath = workDir + "/../Data/Model/dem.tif";

  GDALDataset *img = (GDALDataset *)GDALOpen(demPath.c_str(), GA_ReadOnly);
  if (!img) {
    printf("Can't Open Image!");
    return;
  }
  int bufWidth = img->GetRasterXSize();   //图像宽度
  int bufHeight = img->GetRasterYSize();  //图像高度
  int bandNum = img->GetRasterCount();    //波段数
  if (bandNum != 1) {
    printf("DEM波段数不为1");
    return;
  }
  int depth = GDALGetDataTypeSize(img->GetRasterBand(1)->GetRasterDataType()) /
              8;  //图像深度

  //获取地理坐标信息
  double padfTransform[6];
  if (img->GetGeoTransform(padfTransform) == CE_Failure) {
    printf("获取仿射变换参数失败");
    return;
  }

  double startX = padfTransform[0];
  double dX = padfTransform[1];
  double startY = padfTransform[3];
  double dY = padfTransform[5];

  //申请buf
  size_t imgBufNum = (size_t)bufWidth * bufHeight * bandNum;
  float *imgBuf = new float[imgBufNum];

  //读取
  img->RasterIO(GF_Read, 0, 0, bufWidth, bufHeight, imgBuf, bufWidth, bufHeight,
                GDT_Float32, bandNum, nullptr, bandNum * depth,
                bufWidth * bandNum * depth, depth);

  pointNum = (size_t)bufWidth * bufHeight;
  size_t position_texture_num = pointNum * 5;
  float *position_texture = new float[position_texture_num];

  for (int yi = 0; yi < bufHeight; yi++) {
    for (int xi = 0; xi < bufWidth; xi++) {
      size_t n = (size_t)(bufWidth * 5) * yi + 5 * xi;
      position_texture[n] = dX * xi;
      position_texture[n + 1] = dY * yi;
      size_t m = (size_t)(bufWidth * bandNum) * yi + bandNum * xi;
      position_texture[n + 2] = imgBuf[m];
      position_texture[n + 3] = float(xi) / (bufWidth - 1);
      position_texture[n + 4] = float(yi) / (bufHeight - 1);
    }
  }

  //释放
  delete[] imgBuf;
  imgBuf = nullptr;

  string binPath = workDir + "/../Data/Model/new.bin";
  ofstream binFile(binPath, std::ios::binary);

  binFile.write((char *)position_texture, position_texture_num * sizeof(float));

  size_t vertexBufNum = position_texture_num * sizeof(float);
  binBufNum = binBufNum + vertexBufNum;

  int mod = vertexBufNum % sizeof(uint16_t);
  if (mod != 0) {
    int spaceNum = sizeof(float) - mod;
    char *space = new char[spaceNum];
    binBufNum = binBufNum + sizeof(char) * spaceNum;
    memset(space, 0, sizeof(char) * spaceNum);
    binFile.write(space, sizeof(char) * spaceNum);
    delete[] space;
    space = nullptr;
  }

  indicesNum = (size_t)(bufWidth - 1) * (bufHeight - 1) * 2 * 3;
  uint16_t *indices = new uint16_t[indicesNum];

  for (int yi = 0; yi < bufHeight - 1; yi++) {
    for (int xi = 0; xi < bufWidth - 1; xi++) {
      uint16_t m00 = (uint16_t)(bufWidth * yi + xi);
      uint16_t m01 = (uint16_t)(bufWidth * (yi + 1) + xi);
      uint16_t m11 = (uint16_t)(bufWidth * (yi + 1) + xi + 1);
      uint16_t m10 = (uint16_t)(bufWidth * yi + xi + 1);

      size_t n = (size_t)(bufWidth - 1) * yi + xi;
      indices[n * 6] = m00;
      indices[n * 6 + 1] = m01;
      indices[n * 6 + 2] = m11;
      indices[n * 6 + 3] = m11;
      indices[n * 6 + 4] = m10;
      indices[n * 6 + 5] = m00;
    }
  }

  binFile.write((char *)indices, sizeof(uint16_t) * indicesNum);
  binBufNum = binBufNum + sizeof(uint16_t) * indicesNum;

  delete[] position_texture;
  position_texture = nullptr;

  delete[] indices;
  indices = nullptr;
}

int main() {
  GDALAllRegister();
  CPLSetConfigOption("GDAL_FILENAME_IS_UTF8", "NO");  //支持中文路径

  //设置Proj数据
  std::string projDataPath = getenv("GISBasic");
  projDataPath += "/share/proj";
  CPLSetConfigOption("PROJ_LIB", projDataPath.c_str());

  ordered_json gltf;

  gltf["asset"] = {{"generator", "CL"}, {"version", "2.0"}};

  gltf["scene"] = 0;
  gltf["scenes"] = {{{"nodes", {0}}}};

  gltf["nodes"] = {{{"mesh", 0}}};

  ordered_json positionJson;
  positionJson["POSITION"] = 1;
  positionJson["TEXCOORD_0"] = 2;

  ordered_json primitivesJson;
  primitivesJson = {
      {{"attributes", positionJson}, {"indices", 0}, {"material", 0}}};

  gltf["meshes"] = {{{"primitives", primitivesJson}}};

  ordered_json pbrJson;
  pbrJson["baseColorTexture"]["index"] = 0;

  gltf["materials"] = {{{"pbrMetallicRoughness", pbrJson}}};

  CreateBinFile();

  gltf["textures"] = {{{"sampler", 0}, {"source", 0}}};

  gltf["images"] = {{{"uri", "tex.jpg"}}};

  gltf["samplers"] = {{{"magFilter", 9729},
                       {"minFilter", 9987},
                       {"wrapS", 33648},
                       {"wrapT", 33648}}};

  gltf["buffers"] = {{{"uri", "new.bin"}, {"byteLength", binBufNum}}};

  ordered_json indicesBufferJson;
  indicesBufferJson["buffer"] = 0;
  indicesBufferJson["byteOffset"] = pointNum * 5 * 4;
  indicesBufferJson["byteLength"] = indicesNum * 2;
  indicesBufferJson["target"] = 34963;

  ordered_json positionBufferJson;
  positionBufferJson["buffer"] = 0;
  positionBufferJson["byteStride"] = sizeof(float) * 5;
  positionBufferJson["byteOffset"] = 0;
  positionBufferJson["byteLength"] = pointNum * 5 * 4;
  positionBufferJson["target"] = 34962;

  gltf["bufferViews"] = {indicesBufferJson, positionBufferJson};

  ordered_json indicesAccessors;
  indicesAccessors["bufferView"] = 0;
  indicesAccessors["byteOffset"] = 0;
  indicesAccessors["componentType"] = 5123;
  indicesAccessors["count"] = indicesNum;
  indicesAccessors["type"] = "SCALAR";
  indicesAccessors["max"] = {18719};
  indicesAccessors["min"] = {0};

  ordered_json positionAccessors;
  positionAccessors["bufferView"] = 1;
  positionAccessors["byteOffset"] = 0;
  positionAccessors["componentType"] = 5126;
  positionAccessors["count"] = pointNum;
  positionAccessors["type"] = "VEC3";
  positionAccessors["max"] = {770, 0.0, 1261.151611328125};
  positionAccessors["min"] = {0.0, -2390, 733.5555419921875};

  ordered_json textureAccessors;
  textureAccessors["bufferView"] = 1;
  textureAccessors["byteOffset"] = sizeof(float) * 3;
  textureAccessors["componentType"] = 5126;
  textureAccessors["count"] = pointNum;
  textureAccessors["type"] = "VEC2";
  textureAccessors["max"] = {1, 1};
  textureAccessors["min"] = {0, 0};

  gltf["accessors"] = {indicesAccessors, positionAccessors, textureAccessors};

  string workDir = getenv("GISBasic");
  string jsonFile = workDir + "/../Data/Model/new.gltf";
  ofstream binFile(jsonFile, std::ios::binary);

  std::ofstream outFile(jsonFile);
  outFile << std::setw(4) << gltf << std::endl;
}

File: Python/example-7-3-dem-model.py
import os
import json
import struct
import sys
import numpy as np
from osgeo import gdal


point_num = 0
bin_buf_num = 0
indices_num = 0


def create_bin_file():
    global point_num
    global bin_buf_num
    global indices_num

    work_dir = os.environ["GISBasicRepo"]

    dem_path = os.path.join(work_dir, "Data", "Model", "dem.tif")

    img = gdal.Open(dem_path)

    if img is None:
        print("Can't Open Image!")
        return

    buf_width = img.RasterXSize
    buf_height = img.RasterYSize
    band_num = img.RasterCount

    if band_num != 1:
        print("DEM波段数不为1")
        return

    band = img.GetRasterBand(1)

    geo = img.GetGeoTransform()

    start_x = geo[0]
    dx = geo[1]
    start_y = geo[3]
    dy = geo[5]

    img_buf = band.ReadAsArray().astype(np.float32)

    point_num = buf_width * buf_height

    # xyz + uv
    position_texture = np.zeros(
        (point_num, 5),
        dtype=np.float32
    )

    for yi in range(buf_height):

        y = dy * yi
        v = yi / (buf_height - 1)

        for xi in range(buf_width):

            idx = yi * buf_width + xi

            position_texture[idx, 0] = dx * xi
            position_texture[idx, 1] = y
            position_texture[idx, 2] = img_buf[yi, xi]

            position_texture[idx, 3] = xi / (buf_width - 1)
            position_texture[idx, 4] = v

    bin_path = os.path.join(work_dir, "Data", "Model", "dem.bin")

    with open(bin_path, "wb") as f:

        vertex_bytes = position_texture.tobytes()

        f.write(vertex_bytes)

        vertex_buf_num = len(vertex_bytes)
        bin_buf_num += vertex_buf_num

        # 2字节对齐
        mod = vertex_buf_num % 2

        if mod != 0:

            space_num = 4 - mod

            f.write(bytes(space_num))

            bin_buf_num += space_num

        indices_num = (
            (buf_width - 1)
            * (buf_height - 1)
            * 2
            * 3
        )

        indices = np.zeros(
            indices_num,
            dtype=np.uint16
        )

        for yi in range(buf_height - 1):

            for xi in range(buf_width - 1):

                m00 = buf_width * yi + xi
                m01 = buf_width * (yi + 1) + xi
                m11 = buf_width * (yi + 1) + xi + 1
                m10 = buf_width * yi + xi + 1

                n = yi * (buf_width - 1) + xi

                base = n * 6

                indices[base + 0] = m00
                indices[base + 1] = m01
                indices[base + 2] = m11

                indices[base + 3] = m11
                indices[base + 4] = m10
                indices[base + 5] = m00

        idx_bytes = indices.tobytes()

        f.write(idx_bytes)

        bin_buf_num += len(idx_bytes)

    xyz = position_texture[:, :3]

    return {
        "max_pos": xyz.max(axis=0).tolist(),
        "min_pos": xyz.min(axis=0).tolist(),
        "max_index": int(indices.max())
    }

# 自动设置 PROJ_LIB for Conda (Windows/Linux/macOS) 
def setup_proj_lib():
    if 'CONDA_PREFIX' in os.environ:
        conda_prefix = os.environ['CONDA_PREFIX']
        # Windows
        proj_path = os.path.join(conda_prefix, 'Library', 'share', 'proj')
        if not os.path.exists(proj_path):
            # Unix-like
            proj_path = os.path.join(conda_prefix, 'share', 'proj')
        if os.path.exists(proj_path):       
            os.environ['PROJ_LIB'] = proj_path
            return True
    return False

def main():

    global point_num
    global bin_buf_num
    global indices_num

    gdal.AllRegister()

    setup_proj_lib()

    # 设置 PROJ_LIB 环境变量
    work_dir = os.getenv("GISBasicRepo") 
    if not work_dir:
        print("错误：环境变量 GISBasicRepo 未设置。", file=sys.stderr)
        sys.exit(1)

    stat = create_bin_file()

    gltf = {}

    gltf["asset"] = {
        "generator": "CL",
        "version": "2.0"
    }

    gltf["scene"] = 0

    gltf["scenes"] = [
        {
            "nodes": [0]
        }
    ]

    gltf["nodes"] = [
        {
            "mesh": 0
        }
    ]

    gltf["meshes"] = [
        {
            "primitives": [
                {
                    "attributes": {
                        "POSITION": 1,
                        "TEXCOORD_0": 2
                    },
                    "indices": 0,
                    "material": 0
                }
            ]
        }
    ]

    gltf["materials"] = [
        {
            "pbrMetallicRoughness": {
                "baseColorTexture": {
                    "index": 0
                }
            }
        }
    ]

    gltf["textures"] = [
        {
            "sampler": 0,
            "source": 0
        }
    ]

    gltf["images"] = [
        {
            "uri": "tex.jpg"
        }
    ]

    gltf["samplers"] = [
        {
            "magFilter": 9729,
            "minFilter": 9987,
            "wrapS": 33648,
            "wrapT": 33648
        }
    ]

    gltf["buffers"] = [
        {
            "uri": "dem.bin",
            "byteLength": bin_buf_num
        }
    ]

    vertex_bytes = point_num * 5 * 4

    gltf["bufferViews"] = [

        {
            "buffer": 0,
            "byteOffset": vertex_bytes,
            "byteLength": indices_num * 2,
            "target": 34963
        },

        {
            "buffer": 0,
            "byteStride": 20,
            "byteOffset": 0,
            "byteLength": vertex_bytes,
            "target": 34962
        }

    ]

    gltf["accessors"] = [

        {
            "bufferView": 0,
            "byteOffset": 0,
            "componentType": 5123,
            "count": indices_num,
            "type": "SCALAR",
            "max": [stat["max_index"]],
            "min": [0]
        },

        {
            "bufferView": 1,
            "byteOffset": 0,
            "componentType": 5126,
            "count": point_num,
            "type": "VEC3",
            "max": stat["max_pos"],
            "min": stat["min_pos"]
        },

        {
            "bufferView": 1,
            "byteOffset": 12,
            "componentType": 5126,
            "count": point_num,
            "type": "VEC2",
            "max": [1, 1],
            "min": [0, 0]
        }

    ]

    gltf_path = os.path.join(work_dir, "Data", "Model", "dem.gltf")

    with open(
        gltf_path,
        "w",
        encoding="utf-8"
    ) as f:

        json.dump(
            gltf,
            f,
            indent=4
        )

    print("完成")
    print("顶点:", point_num)
    print("索引:", indices_num)


if __name__ == "__main__":
    main()